#출처

https://velog.io/@ezoo0422/Python-pycaret%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%98%EC%97%AC-%EB%AA%A8%EB%8D%B8-%EC%84%A0%EC%A0%95%ED%95%98%EA%B8%B0

#라이브러리 설치하기

In [1]:
!pip install pycaret

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 320 kB 30.1 MB/s 
     |████████████████████████████████| 88 kB 10.3 MB/s 
     |████████████████████████████████| 6.8 MB 73.6 MB/s 
     |████████████████████████████████| 56 kB 6.5 MB/s 
     |████████████████████████████████| 262 kB 86.1 MB/s 
     |████████████████████████████████| 10.4 MB 68.8 MB/s 
     |████████████████████████████████| 1.3 MB 85.4 MB/s 
     |████████████████████████████████| 1.7 MB 79.7 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 636 kB 67.1 MB/s 
     |████████████████████████████████| 167 kB 84.0 MB/s 
     |████████████████████████████████| 17.0 MB 74.3 MB/s 
     |████████████████████████████████| 25.9 MB 1.3 MB/s 
     |███████████████████████████████

In [7]:
import numpy

In [3]:
!pip install markupsafe==2.0.1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Attempting uninstall: markupsafe
    Found existing installation: MarkupSafe 2.1.1
    Uninstalling MarkupSafe-2.1.1:
      Successfully uninstalled MarkupSafe-2.1.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pandas-profiling 3.2.0 requires markupsafe~=2.1.1, but you have markupsafe 2.0.1 which is incompatible.


#데이터 가져오기

In [4]:
from pycaret.datasets import get_data
dataset = get_data('credit')

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0      20000    2          2         1   24      2      2     -1     -1   
1      90000    2          2         2   34      0      0      0      0   
2      50000    2          2         1   37      0      0      0      0   
3      50000    1          2         1   57     -1      0     -1      0   
4      50000    1          1         2   37      0      0      0      0   

   PAY_5  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0     -2  ...        0.0        0.0        0.0       0.0     689.0       0.0   
1      0  ...    14331.0    14948.0    15549.0    1518.0    1500.0    1000.0   
2      0  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
3      0  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   
4      0  ...    19394.0    19619.0    20024.0    2500.0    1815.0     657.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default  
0       0.0       0.0       0.0        1  
1    1000.0    1000.0    5000.0        0  
2    1100.0    1069.0    1000.0        0  
3    9000.0     689.0     679.0        0  
4    1000.0    1000.0     800.0        0  

[5 rows x 24 columns]

train/test 데이터로 데이터를 분리

In [5]:
#frac: 특정 비율의 표본을 무작위로 추출하기
#random_state=1 이라고 하면 바로 이 random 함수의 seed 값을 고정시키기 때문에 여러번 수행하더라도 같은 레코드를 추출합니다.
train = dataset.sample(frac=0.95, random_state=786)
#drop = 삭제
test = dataset.drop(train.index)
train.reset_index(inplace=True, drop=True)
test.reset_index(inplace=True, drop=True)

#데이터 설정

In [13]:
from pycaret.classification import *

exp_clf = setup(data = train, target = 'default', session_id=123)

#data 설정값이 다음과 같이 출력됩니다.
#exp_clf = setup(data = train, target = 'default', session_id=123)이런 순서
#마지막에 enter를 처야함
#session_id: random_state와 같은 개념으로 같은 결과가 나올 수 있게 seed를 고정
#data: train 데이터를 입력
#target = target 변수 이름을 입력

Description             Value
0                               session_id               123
1                                   Target           default
2                              Target Type            Binary
3                            Label Encoded              None
4                            Original Data       (22800, 24)
5                           Missing Values             False
6                         Numeric Features                14
7                     Categorical Features                 9
8                         Ordinal Features             False
9                High Cardinality Features             False
10                 High Cardinality Method              None
11                   Transformed Train Set       (15959, 88)
12                    Transformed Test Set        (6841, 88)
13                      Shuffle Train-Test              True
14                     Stratify Train-Test             False
15                          Fold Generator   StratifiedKFold
16                             Fold Number                10
17                                CPU Jobs                -1
18                                 Use GPU             False
19                          Log Experiment             False
20                         Experiment Name  clf-default-name
21                                     USI              2e04
22                         Imputation Type            simple
23          Iterative Imputation Iteration              None
24                         Numeric Imputer              mean
25      Iterative Imputation Numeric Model              None
26                     Categorical Imputer          constant
27  Iterative Imputation Categorical Model              None
28           Unknown Categoricals Handling    least_frequent
29                               Normalize             False
30                        Normalize Method              None
31                          Transformation             False
32                   Transformation Method              None
33                                     PCA             False
34                              PCA Method              None
35                          PCA Components              None
36                     Ignore Low Variance             False
37                     Combine Rare Levels             False
38                    Rare Level Threshold              None
39                         Numeric Binning             False
40                         Remove Outliers             False
41                      Outliers Threshold              None
42                Remove Multicollinearity             False
43             Multicollinearity Threshold              None
44             Remove Perfect Collinearity              True
45                              Clustering             False
46                    Clustering Iteration              None
47                     Polynomial Features             False
48                       Polynomial Degree              None
49                    Trignometry Features             False
50                    Polynomial Threshold              None
51                          Group Features             False
52                       Feature Selection             False
53                Feature Selection Method           classic
54            Features Selection Threshold              None
55                     Feature Interaction             False
56                           Feature Ratio             False
57                   Interaction Threshold              None
58                           Fix Imbalance             False
59                    Fix Imbalance Method             SMOTE

INFO:logs:create_model_container: 0
INFO:logs:master_model_container: 0
INFO:logs:display_container: 1
INFO:logs:Pipeline(memory=None,
         steps=[('dtypes',
                 DataTypes_Auto_infer(categorical_features=[],
                                      display_types=True, features_todrop=[],
                                      id_columns=[],
                                      ml_usecase='classification',
                                      numerical_features=[], target='default',
                                      time_features=[])),
                ('imputer',
                 Simple_Imputer(categorical_strategy='not_available',
                                fill_value_categorical=None,
                                fill_value_numerical=None,
                                numeric_stra...
                ('scaling', 'passthrough'), ('P_transform', 'passthrough'),
                ('binn', 'passthrough'), ('rem_outliers', 'passthrough'),
                ('cluste

#모델 비교

compare_models(): 다양한 모델 적합 후 성능 비교

+ fold: cross_validation의 fold를 지정 (default = 10)

+ sort: 정렬기준 지표 설정

+ n_select: 상위 n개의 모델 결과만 출력

In [9]:
best_model = compare_models()

Model  Accuracy     AUC  Recall   Prec.  \
ridge                    Ridge Classifier    0.8254  0.0000  0.3637  0.6913   
lda          Linear Discriminant Analysis    0.8247  0.7634  0.3755  0.6794   
gbc          Gradient Boosting Classifier    0.8226  0.7789  0.3551  0.6806   
ada                  Ada Boost Classifier    0.8221  0.7697  0.3505  0.6811   
lightgbm  Light Gradient Boosting Machine    0.8210  0.7750  0.3609  0.6679   
rf               Random Forest Classifier    0.8199  0.7598  0.3663  0.6601   
et                 Extra Trees Classifier    0.8092  0.7377  0.3677  0.6047   
lr                    Logistic Regression    0.7814  0.6410  0.0003  0.1000   
dummy                    Dummy Classifier    0.7814  0.5000  0.0000  0.0000   
knn                K Neighbors Classifier    0.7547  0.5939  0.1763  0.3719   
dt               Decision Tree Classifier    0.7293  0.6147  0.4104  0.3878   
svm                   SVM - Linear Kernel    0.7277  0.0000  0.1017  0.1671   
qda       Quadratic Discriminant Analysis    0.5098  0.5473  0.6141  0.2472   
nb                            Naive Bayes    0.3760  0.6442  0.8845  0.2441   

              F1   Kappa     MCC  TT (Sec)  
ridge     0.4764  0.3836  0.4122     0.031  
lda       0.4835  0.3884  0.4132     0.154  
gbc       0.4664  0.3725  0.4010     4.110  
ada       0.4626  0.3690  0.3983     0.888  
lightgbm  0.4683  0.3721  0.3977     0.282  
rf        0.4707  0.3727  0.3965     2.064  
et        0.4571  0.3497  0.3657     1.507  
lr        0.0006  0.0003  0.0034     0.533  
dummy     0.0000  0.0000  0.0000     0.013  
knn       0.2388  0.1145  0.1259     0.495  
dt        0.3986  0.2242  0.2245     0.268  
svm       0.0984  0.0067  0.0075     0.211  
qda       0.3488  0.0600  0.0805     0.077  
nb        0.3826  0.0608  0.1207     0.032

INFO:logs:create_model_container: 14
INFO:logs:master_model_container: 14
INFO:logs:display_container: 2
INFO:logs:RidgeClassifier(alpha=1.0, class_weight=None, copy_X=True, fit_intercept=True,
                max_iter=None, normalize=False, random_state=123, solver='auto',
                tol=0.001)
INFO:logs:compare_models() succesfully completed......................................


#모델 적합

앞에서는 다양한 모델의 적합 결과를 확인하였는데, 이번에는 하나의 모델의 적합 결과를 보는 방법

폴드가 10개라 0~9까지 출력

fold: cross_validation의 fold 지정 (default = 10)

In [10]:
rf = create_model('rf')

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
Fold                                                          
0       0.8133  0.7673  0.3610  0.6269  0.4582  0.3551  0.3749
1       0.8239  0.7615  0.3782  0.6735  0.4844  0.3882  0.4117
2       0.8258  0.7708  0.3467  0.7076  0.4654  0.3756  0.4098
3       0.8177  0.7605  0.3725  0.6436  0.4719  0.3710  0.3913
4       0.8208  0.7642  0.3725  0.6599  0.4762  0.3780  0.4006
5       0.8283  0.7638  0.3954  0.6866  0.5018  0.4070  0.4297
6       0.8127  0.7647  0.3582  0.6250  0.4554  0.3522  0.3721
7       0.8283  0.7390  0.3553  0.7168  0.4751  0.3861  0.4202
8       0.8108  0.7496  0.3610  0.6146  0.4549  0.3496  0.3678
9       0.8176  0.7565  0.3621  0.6462  0.4641  0.3645  0.3867
Mean    0.8199  0.7598  0.3663  0.6601  0.4707  0.3727  0.3965
Std     0.0062  0.0089  0.0131  0.0335  0.0139  0.0172  0.0202

INFO:logs:create_model_container: 15
INFO:logs:master_model_container: 15
INFO:logs:display_container: 3
INFO:logs:RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=-1, oob_score=False, random_state=123, verbose=0,
                       warm_start=False)
INFO:logs:create_model() succesfully completed......................................


#하이퍼파라미터 튜닝

tune_model() 함수를 사용해서 모델의 하이퍼파라미터 튜닝을 진행

In [11]:
turned_rf = tune_model(rf)

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
Fold                                                          
0       0.8158  0.7508  0.3181  0.6647  0.4302  0.3363  0.3689
1       0.8283  0.7675  0.3295  0.7419  0.4563  0.3719  0.4152
2       0.8139  0.7337  0.3181  0.6529  0.4277  0.3321  0.3628
3       0.8246  0.7588  0.3095  0.7347  0.4355  0.3514  0.3976
4       0.8170  0.7567  0.3438  0.6557  0.4511  0.3539  0.3805
5       0.8258  0.7506  0.3324  0.7205  0.4549  0.3676  0.4067
6       0.8170  0.7530  0.3324  0.6629  0.4427  0.3474  0.3771
7       0.8221  0.7507  0.3381  0.6901  0.4538  0.3621  0.3951
8       0.8177  0.7201  0.2980  0.6933  0.4168  0.3286  0.3699
9       0.8207  0.7484  0.3132  0.6987  0.4325  0.3439  0.3831
Mean    0.8203  0.7490  0.3233  0.6915  0.4402  0.3495  0.3857
Std     0.0045  0.0126  0.0135  0.0310  0.0129  0.0140  0.0165

INFO:logs:create_model_container: 16
INFO:logs:master_model_container: 16
INFO:logs:display_container: 4
INFO:logs:RandomForestClassifier(bootstrap=False, ccp_alpha=0.0, class_weight={},
                       criterion='entropy', max_depth=5, max_features=1.0,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0002, min_impurity_split=None,
                       min_samples_leaf=5, min_samples_split=10,
                       min_weight_fraction_leaf=0.0, n_estimators=150,
                       n_jobs=-1, oob_score=False, random_state=123, verbose=0,
                       warm_start=False)
INFO:logs:tune_model() succesfully completed......................................


In [15]:
turned_rf

RandomForestClassifier(bootstrap=False, ccp_alpha=0.0, class_weight={},
                       criterion='entropy', max_depth=5, max_features=1.0,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0002, min_impurity_split=None,
                       min_samples_leaf=5, min_samples_split=10,
                       min_weight_fraction_leaf=0.0, n_estimators=150,
                       n_jobs=-1, oob_score=False, random_state=123, verbose=0,
                       warm_start=False)

#Blending

blend_models() 함수를 사용하면 여러 모델들을 혼합하여 새로운 모델을 생성합니다.

모델을 하나씩 생성해서 blend해도 되고 compare_model을 사용하여 생성한 모델을 사용해서 blend할 수 있습니다.

In [12]:
#방법 1 내가 지정한 2개만 blending

df = create_model('dt')
rf = create_model('rf')

blender_2 = blend_models(estimator_list = [df,rf])

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
Fold                                                          
0       0.7343  0.7410  0.4327  0.4005  0.4160  0.2444  0.2447
1       0.7337  0.7428  0.4413  0.4010  0.4202  0.2479  0.2483
2       0.7431  0.7477  0.4241  0.4146  0.4193  0.2544  0.2544
3       0.7274  0.7414  0.4155  0.3856  0.4000  0.2240  0.2242
4       0.7187  0.7317  0.4040  0.3691  0.3858  0.2038  0.2042
5       0.7193  0.7311  0.3926  0.3673  0.3795  0.1984  0.1986
6       0.7206  0.7400  0.4212  0.3760  0.3973  0.2162  0.2168
7       0.7331  0.7168  0.3610  0.3830  0.3717  0.2024  0.2026
8       0.7206  0.7236  0.3983  0.3707  0.3840  0.2036  0.2038
9       0.7442  0.7390  0.4195  0.4148  0.4171  0.2533  0.2533
Mean    0.7295  0.7355  0.4110  0.3883  0.3991  0.2248  0.2251
Std     0.0092  0.0091  0.0219  0.0174  0.0173  0.0218  0.0218

INFO:logs:create_model_container: 19
INFO:logs:master_model_container: 19
INFO:logs:display_container: 7
INFO:logs:VotingClassifier(estimators=[('dt',
                              DecisionTreeClassifier(ccp_alpha=0.0,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features=None,
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                   

In [16]:
# 방법 2 n_select로 뽑은 최고성능 5개를 섞기
#최고로 높은 accuracy를 가진게 ridge로 0.8254을 가지는데 섞으니까 오히려 정확도가 떨어짐
best_model_5 = compare_models(n_select=5)
blender_5 = blend_models(best_model_5)

Accuracy  AUC  Recall   Prec.      F1   Kappa     MCC
Fold                                                       
0       0.8170  0.0  0.3582  0.6477  0.4613  0.3619  0.3849
1       0.8308  0.0  0.3954  0.7005  0.5055  0.4128  0.4374
2       0.8227  0.0  0.3438  0.6897  0.4589  0.3668  0.3986
3       0.8277  0.0  0.3668  0.7033  0.4821  0.3908  0.4206
4       0.8208  0.0  0.3582  0.6684  0.4664  0.3703  0.3964
5       0.8333  0.0  0.4011  0.7107  0.5128  0.4215  0.4466
6       0.8227  0.0  0.3696  0.6719  0.4769  0.3808  0.4055
7       0.8283  0.0  0.3524  0.7193  0.4731  0.3846  0.4196
8       0.8183  0.0  0.3438  0.6630  0.4528  0.3568  0.3844
9       0.8257  0.0  0.3793  0.6804  0.4871  0.3921  0.4165
Mean    0.8247  0.0  0.3669  0.6855  0.4777  0.3838  0.4111
Std     0.0051  0.0  0.0189  0.0219  0.0187  0.0202  0.0199

INFO:logs:create_model_container: 15
INFO:logs:master_model_container: 15
INFO:logs:display_container: 3
INFO:logs:VotingClassifier(estimators=[('ridge',
                              RidgeClassifier(alpha=1.0, class_weight=None,
                                              copy_X=True, fit_intercept=True,
                                              max_iter=None, normalize=False,
                                              random_state=123, solver='auto',
                                              tol=0.001)),
                             ('lda',
                              LinearDiscriminantAnalysis(n_components=None,
                                                         priors=None,
                                                         shrinkage=None,
                                                         solver='svd',
                                                         store_covariance=False,
                                                         tol=0.0001

#예측

finalize_model() 함수로 모델을 설정하면, cross_validation을 사용하여 적합한 모델을 전체 데이터로 마지막으로 학습을 합니다.

 마지막 모델을 설정한 후에 predict_model()을 통해 예측을 합니다.

In [17]:
final_model = finalize_model(blender_5)
prediction = predict_model(final_model, data = test)

INFO:logs:Initializing finalize_model()
INFO:logs:finalize_model(estimator=VotingClassifier(estimators=[('ridge',
                              RidgeClassifier(alpha=1.0, class_weight=None,
                                              copy_X=True, fit_intercept=True,
                                              max_iter=None, normalize=False,
                                              random_state=123, solver='auto',
                                              tol=0.001)),
                             ('lda',
                              LinearDiscriminantAnalysis(n_components=None,
                                                         priors=None,
                                                         shrinkage=None,
                                                         solver='svd',
                                                         store_covariance=False,
                                                         tol=0.0001)),
                             ('gbc',

Model  Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
0  Voting Classifier     0.815  0.6559  0.3726  0.6323  0.4689  0.3658  0.3846

finalize_model(): 최종 모델로 설정 후 마지막 학습 진행

predict_model(): 예측 결과를 'Label' 변수에 저장

최종 0.815의 정확도가 나온다. 모델을 섞으니 더 떨어지네 ㅋㅋㅋ